### ISSUE TITLE
deadCodeInjection: true causes error SyntaxError: function statement not allowed

### ISSUE THREAD

**Issue by ddzobov** [javascript-obfuscator/javascript-obfuscator#499]
When deadCodeInjection enabled:
```
if (_0x5abbc0[_0x58af('0x18a', 'Mg$V')](_0x5abbc0[_0x58af('0x18b', 'TB25')], _0x5abbc0[_0x58af('0x18c', 'gyn0')])) {
                throw new TypeError(_0x4c3a6f[_0x58af('0x18d', '3dd^')]);
            } else {
----->       function _0x556570(_0x2b416c, _0x3ba7ab, _0x5148c6) {
                    var _0x2b0d32 = _0x4c3a6f[_0x58af('0x18e', 'WFsi')][_0x58af('0x18f', 'VjCx')]('|');
                    var _0x21f712 = 0x0;
                    while (!![]) {
                        switch (_0x2b0d32[_0x21f712++]) {
```

I think that it is caused by 'use strict' statement

**Comment by sanex3339** [comment]
Need an example to reproduce. It is the global scope?

**Comment by ddzobov** [comment]
> Need an example to reproduce. It is the global scope?

Can i share it to you privately? Can you write to email ddzobov@gmail.com?

Or i will try to extract part of problem code later.

**Comment by sanex3339** [comment]
I will write to yu but it is better to find the more small part with the problem.

**Comment by ddzobov** [comment]
Source:
```
export default class App {
    ready(callback) {
        let startupInterval = setInterval(() => {
            clearInterval(startupInterval);

            callback();
        }, 100);
    }
}

new App().ready(() => {
    console.log('OK');
});
```

Build result:
```
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "/";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ({

/***/ "./resources/js/loader.js":
/*!********************************!*\
  !*** ./resources/js/loader.js ***!
  \********************************/
/*! exports provided: default */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "default", function() { return App; });
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var App =
/*#__PURE__*/
function () {
  function App() {
    _classCallCheck(this, App);
  }

  _createClass(App, [{
    key: "ready",
    value: function ready(callback) {
      var startupInterval = setInterval(function () {
        clearInterval(startupInterval);
        callback();
      }, 100);
    }
  }]);

  return App;
}();


new App().ready(function () {
  console.log('OK');
});

/***/ }),

/***/ 0:
/*!**************************************!*\
  !*** multi ./resources/js/loader.js ***!
  \**************************************/
/*! no static exports found */
/***/ (function(module, exports, __webpack_require__) {

module.exports = __webpack_require__(/*! /Users/zobov/Desktop/projects/test/resources/js/loader.js */"./resources/js/loader.js");


/***/ })

/******/ });
```

Config:
```
{
    deadCodeInjection: true
}
```

**Comment by ddzobov** [comment]
Now i see that it is floating error.

Some builds without config change have ran successfully.

**Comment by ddzobov** [comment]
Success 1:
```
'./resources/js/loader.js': function(_0x3542cc, _0x484ef9, _0x38b909) {
        'use strict';
        _0x38b909['r'](_0x484ef9);
        _0x38b909['d'](_0x484ef9, _0x5b8f('0x9'), function() {
            if ('bwPKu' === 'gGGDV') {
                function _0x52eb27() {
                    _0x20d006(this, _0x52eb27);
                }
                _0x4af099(_0x52eb27, [{
                    'key': _0x5b8f('0xf'),
                    'value': function ready(_0x1eb30c) {
                        var _0x48b5ee = setInterval(function() {
                            clearInterval(_0x48b5ee);
                            _0x1eb30c();
                        }, 0x64);
                    }
                }]);
                return _0x52eb27;
            } else {
                return _0x190d0a;
            }
        });

        function _0x20d006(_0x1aa734, _0x4df568) {
            if (_0x5b8f('0x10') !== _0x5b8f('0x11')) {
                if (!(_0x1aa734 instanceof _0x4df568)) {
                    throw new TypeError(_0x5b8f('0x5'));
                }
            } else {
                Object['defineProperty'](exports, name, {
                    'enumerable': !![],
                    'get': getter
                });
            }
        }

        function _0xdd6501(_0x59e419, _0x3337b6) {
            for (var _0x332bff = 0x0; _0x332bff < _0x3337b6[_0x5b8f('0x12')]; _0x332bff++) {
                var _0x5af9f9 = _0x3337b6[_0x332bff];
                _0x5af9f9[_0x5b8f('0x13')] = _0x5af9f9[_0x5b8f('0x13')] || ![];
                _0x5af9f9[_0x5b8f('0x14')] = !![];
                if ('value' in _0x5af9f9) _0x5af9f9['writable'] = !![];
                Object[_0x5b8f('0x1')](_0x59e419, _0x5af9f9[_0x5b8f('0x15')], _0x5af9f9);
            }
        }

        function _0x4af099(_0x20c192, _0x140e36, _0x3590ab) {
            if (_0x5b8f('0x16') !== _0x5b8f('0x16')) {
                return _0x3542cc;
            } else {
                if (_0x140e36) _0xdd6501(_0x20c192[_0x5b8f('0xc')], _0x140e36);
                if (_0x3590ab) _0xdd6501(_0x20c192, _0x3590ab);
                return _0x20c192;
            }
        }
        var _0x190d0a = function() {
            if (_0x5b8f('0x17') !== _0x5b8f('0x17')) {
                if (mode & 0x1) value = _0x38b909(value);
                if (mode & 0x8) return value;
                if (mode & 0x4 && typeof value === _0x5b8f('0x7') && value && value[_0x5b8f('0x8')]) return value;
                var _0x26d36a = Object[_0x5b8f('0x18')](null);
                _0x38b909['r'](_0x26d36a);
                Object[_0x5b8f('0x1')](_0x26d36a, _0x5b8f('0x9'), {
                    'enumerable': !![],
                    'value': value
                });
                if (mode & 0x2 && typeof value != _0x5b8f('0x19'))
                    for (var _0x3b2161 in value) _0x38b909['d'](_0x26d36a, _0x3b2161, function(_0x3b2161) {
                        return value[_0x3b2161];
                    } [_0x5b8f('0xa')](null, _0x3b2161));
                return _0x26d36a;
            } else {
                function _0x190d0a() {
                    _0x20d006(this, _0x190d0a);
                }
                _0x4af099(_0x190d0a, [{
                    'key': _0x5b8f('0xf'),
                    'value': function ready(_0x1f97a8) {
                        if ('AHoMO' === _0x5b8f('0x1a')) {
                            var _0x221ec4 = setInterval(function() {
                                clearInterval(_0x221ec4);
                                _0x1f97a8();
                            }, 0x64);
                        } else {
                            var _0x2afaf1 = setInterval(function() {
                                clearInterval(_0x2afaf1);
                                _0x1f97a8();
                            }, 0x64);
                        }
                    }
                }]);
                return _0x190d0a;
            }
        }();
        new _0x190d0a()[_0x5b8f('0xf')](function() {
            console['log']('OK');
        });
    }
```

Success 2:
```
'./resources/js/loader.js': function(_0x2fe73c, _0x191e15, _0x2c444f) {
        'use strict';
        _0x2c444f['r'](_0x191e15);
        _0x2c444f['d'](_0x191e15, _0xb20c('0x8'), function() {
            return _0x57bb6a;
        });

        function _0x31953f(_0x942d41, _0x52cda1) {
            if (_0xb20c('0x13') !== 'UxPHi') {
                if (!(_0x942d41 instanceof _0x52cda1)) {
                    throw new TypeError(_0xb20c('0x14'));
                }
            } else {
                return _0x2fe73c['default'];
            }
        }

        function _0x69676f(_0x590d81, _0x29f273) {
            for (var _0x1b8850 = 0x0; _0x1b8850 < _0x29f273[_0xb20c('0x15')]; _0x1b8850++) {
                var _0x252d49 = _0x29f273[_0x1b8850];
                _0x252d49[_0xb20c('0xf')] = _0x252d49[_0xb20c('0xf')] || ![];
                _0x252d49['configurable'] = !![];
                if (_0xb20c('0x10') in _0x252d49) _0x252d49[_0xb20c('0x11')] = !![];
                Object['defineProperty'](_0x590d81, _0x252d49['key'], _0x252d49);
            }
        }

        function _0xd65b02(_0x4721b6, _0x33356b, _0x286daa) {
            if (_0x33356b) _0x69676f(_0x4721b6[_0xb20c('0xd')], _0x33356b);
            if (_0x286daa) _0x69676f(_0x4721b6, _0x286daa);
            return _0x4721b6;
        }
        var _0x57bb6a = function() {
            if (_0xb20c('0x16') !== 'lNWiR') {
----->     function _0x57bb6a() {
                    if (_0xb20c('0x17') === 'ysUSS') {
                        return _0x57bb6a;
                    } else {
                        _0x31953f(this, _0x57bb6a);
                    }
                }
                _0xd65b02(_0x57bb6a, [{
                    'key': 'ready',
                    'value': function ready(_0x5a4ea6) {
                        var _0x24eda3 = setInterval(function() {
                            clearInterval(_0x24eda3);
                            _0x5a4ea6();
                        }, 0x64);
                    }
                }]);
                return _0x57bb6a;
            } else {
                if (protoProps) _0x69676f(Constructor[_0xb20c('0xd')], protoProps);
                if (staticProps) _0x69676f(Constructor, staticProps);
                return Constructor;
            }
        }();
        new _0x57bb6a()['ready'](function() {
            if (_0xb20c('0x18') !== _0xb20c('0x18')) {
                throw new TypeError(_0xb20c('0x14'));
            } else {
                console[_0xb20c('0x19')]('OK');
            }
        });
    }
```

Error:
```
'./resources/js/loader.js': function(_0x353f28, _0x589f86, _0x12bf0b) {
        'use strict';
        _0x12bf0b['r'](_0x589f86);
        _0x12bf0b['d'](_0x589f86, _0x2ea3('0xb'), function() {
            if (_0x2ea3('0x12') === _0x2ea3('0x12')) {
                return _0x6cc646;
            } else {
                var _0x139f53 = setInterval(function() {
                    clearInterval(_0x139f53);
                    callback();
                }, 0x64);
            }
        });

        function _0x321a65(_0xfa959a, _0x2e0924) {
            if (_0x2ea3('0x13') !== _0x2ea3('0x14')) {
                if (!(_0xfa959a instanceof _0x2e0924)) {
                    throw new TypeError(_0x2ea3('0x15'));
                }
            } else {
                return Object['prototype'][_0x2ea3('0x11')]['call'](object, property);
            }
        }

        function _0x49be38(_0x4ee97a, _0x59be1b) {
            for (var _0x322e25 = 0x0; _0x322e25 < _0x59be1b['length']; _0x322e25++) {
                if (_0x2ea3('0x16') !== _0x2ea3('0x16')) {
                    if (!(instance instanceof Constructor)) {
                        throw new TypeError('Cannot\x20call\x20a\x20class\x20as\x20a\x20function');
                    }
                } else {
                    var _0x4187c6 = _0x59be1b[_0x322e25];
                    _0x4187c6[_0x2ea3('0x17')] = _0x4187c6[_0x2ea3('0x17')] || ![];
                    _0x4187c6[_0x2ea3('0x18')] = !![];
                    if (_0x2ea3('0x19') in _0x4187c6) _0x4187c6[_0x2ea3('0x1a')] = !![];
                    Object[_0x2ea3('0x3')](_0x4ee97a, _0x4187c6['key'], _0x4187c6);
                }
            }
        }

        function _0x592c34(_0x52c8a1, _0x44de75, _0x5d1b0f) {
            if (_0x2ea3('0x1b') === _0x2ea3('0x1b')) {
                if (_0x44de75) _0x49be38(_0x52c8a1[_0x2ea3('0xe')], _0x44de75);
                if (_0x5d1b0f) _0x49be38(_0x52c8a1, _0x5d1b0f);
                return _0x52c8a1;
            } else {
                return value[key];
            }
        }
        var _0x6cc646 = function() {
            function _0x6cc646() {
                _0x321a65(this, _0x6cc646);
            }
            _0x592c34(_0x6cc646, [{
                'key': 'ready',
                'value': function ready(_0x693708) {
                    if ('YJTmd' === _0x2ea3('0x1c')) {
                        var _0x379e7e = setInterval(function() {
                            if ('rtPfx' === _0x2ea3('0x1d')) {
                                clearInterval(_0x379e7e);
                                _0x693708();
                            } else {
                                if (!_0x12bf0b['o'](exports, name)) {
                                    Object['defineProperty'](exports, name, {
                                        'enumerable': !![],
                                        'get': getter
                                    });
                                }
                            }
                        }, 0x64);
                    } else {
                        _0x353f28['exports'] = _0x12bf0b(_0x2ea3('0x1e'));
                    }
                }
            }]);
            return _0x6cc646;
        }();
        new _0x6cc646()['ready'](function() {
            console[_0x2ea3('0x1f')]('OK');
        });
    }
```

**Comment by sanex3339** [comment]
Nice. Will lok into that in the next few days

**Comment by sanex3339** [comment]
Need more info, how to reproduce it

**Comment by ddzobov** [comment]
Hm, i can repeat it only in Duktape. In strict mode required to place all function definitions before all statements. Obsuscator sometimes placing function definitions into statements.

When function placed here, it causes error:
```
        var _0x57bb6a = function() {
            if (_0xb20c('0x16') !== 'lNWiR') {
----->          function _0x57bb6a() {
                    if (_0xb20c('0x17') === 'ysUSS') {
                        return _0x57bb6a;
                    } else {
                        _0x31953f(this, _0x57bb6a);
                    }
                }
```

This variant OK:
```
	var _0x57bb6a = function() {
		function _0x57bb6a() {
			if (_0xb20c('0x17') === 'ysUSS') {
				return _0x57bb6a;
			} else {
				_0x31953f(this, _0x57bb6a);
			}
		}
		if (_0xb20c('0x16') !== 'lNWiR') {
```

**Comment by da411d** [comment]
I'm not expert in JavaScript specification, but is some environments it is denied to declare functions in `if` statements because of unpredictability.
Source: https://stackoverflow.com/a/10069457/5698975

I suggest to replace function definition to function expression:
```
var _0x57bb6a = function() {
  if (_0xb20c('0x16') !== 'lNWiR') {
    // function _0x57bb6a() {
    let _0x57bb6a = function _0x57bb6a() {
      if (_0xb20c('0x17') === 'ysUSS') {
        return _0x57bb6a;
      } else {
        _0x31953f(this, _0x57bb6a);
      }
    };
    // ...
  }
};
```

**Comment by sanex3339** [comment]
https://github.com/javascript-obfuscator/javascript-obfuscator/pull/655

Collected block statements with `FunctionDeclaration` nodes will be ignored.

**Comment by sanex3339** [comment]
Merged. Will be released soon

**Comment by sanex3339** [comment]
Released as `1.3.0`
